### PR TITLE
[PORT] DB Achievements

### PR DESCRIPTION
**Pull request by Ivniinvi**
## About The Pull Request
Ports DB achievements from TG.
Ports the following:
tgstation/tgstation#47058
tgstation/tgstation#47617
tgstation/tgstation#48007
tgstation/tgstation#48146
And the relevant parts of:
tgstation/tgstation#51255

**This will require migration using the enclosed tool**

## Why It's Good For The Game
We no longer rely on the BYOND hub for achievements, instead using the DB.

## Changelog
:cl: Ivniinvi, AMonkeyThatCodes, Qustinnus, Willox, AnturK, SpaceManiac
server: Achievements are now handled by the DB instead of the BYOND hub
add: The verb check-achievements in the OOC tab can be used to see your achievements
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
based

**Comment by qwertyquerty** [comment]
Bring in all the existing medals into this system

**Comment by Ivniinvi** [comment]
> Bring in all the existing medals into this system

It has everything except G-G-Ghosts

**Comment by Ivniinvi** [comment]
I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.

**Comment by qwertyquerty** [comment]
Alright, so I think we need a way to award the users who currently have medals their respective achievements.

**Comment by Ivniinvi** [comment]
> Alright, so I think we need a way to award the users who currently have medals their respective achievements.

Yes, that's what HubMigrator does.

**Comment by KubeRoot** [comment]
> I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.

Maybe edit the PR description to remove the mention of not having that one

**Comment by Ivniinvi** [comment]
> > I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.
> 
> Maybe edit the PR description to remove the mention of not having that one

Oop, thanks, updated.

**Review by qwertyquerty — DISMISSED** [review]
Remove your config changes

**Comment by Ivniinvi** [comment]
image

Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)

File: tools/HubMigrator/HubMigrator.dm [hunk 1]
```diff
@@ -0,0 +1,183 @@
+//Misc Medal hub IDs
+#define MEDAL_METEOR 			"Your Life Before Your Eyes"
```
Comment (Crossedfall) [inline tools/HubMigrator/HubMigrator.dm:2]: I only did a brief check, but I found 10 that didn't correctly match up. Should go back through this list and make sure it's all correct.

File: code/datums/achievements/boss_achievements.dm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+/datum/award/achievement/boss/tendril_exterminator
+	name = "Tendril Exterminator"
+	desc = "Watch your step"
+	database_id = BOSS_MEDAL_TENDRIL
+
+/datum/award/achievement/boss/boss_killer
+	name = "Boss Killer"
+	desc = "You've come a long ways from asking how to switch hands."
+	database_id = "Boss Killer"
```
Comment (Crossedfall) [inline code/datums/achievements/boss_achievements.dm]:
```suggestion
	database_id = BOSS_MEDAL_ANY
```

**Review by Crossedfall — CHANGES_REQUESTED** [review]

**Comment by Crossedfall** [comment]
> image
> 
> Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)

I haven't approved anything yet, I just told you that I would only approve the stuff I had ownership of.

**Comment by Ivniinvi** [comment]
> > Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)
> 
> I haven't approved anything yet, I just told you that I would only approve the stuff I had ownership of.

Oops, sorry, misinterpreted that

**Comment by Ivniinvi** [comment]
Right, I've addressed the concerns presented.

Various inconsistencies:
The "Exterminator" medal exists in the code and on the hub, but no one has it.
The "Crusher" medal exists in the code but not on the hub.
The "Forgotten Star" medal exists in the code but not on the hub.
The "Avada Kedavra!" medal exists on the hub but not in the code, and no one has it.

As such, none of these four have been included in HubMigrator.dm

File: tools/HubMigrator/HubMigrator.dm [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	for(var/key in giant_list_of_ckeys)
+		world << "Generating entries for [key] [i]/[giant_list_of_ckeys.len]"
+		var/keyv = ckey(key) //Checkinf if you don't have any manually entered drop tables; juniors on your hub is good idea.
+		var/list/values = list()
+		for(var/cheevo in medal_data[keyv])
+			values += "('[keyv]','[cheevo]',1, '[medal_data[keyv][cheevo]]')"
+		for(var/score in scores_data[keyv])
+			values += "('[keyv]','[score]',[scores_data[keyv][score]],now())"
+		if(values.len)
+			var/list/keyline = list("INSERT INTO [ach](ckey,achievement_key,value,last_updated) VALUES")
```
Comment (Crossedfall) [inline tools/HubMigrator/HubMigrator.dm:161]: last_updated should be when the achievement was unlocked (which is already part of the scraped data).  Currently, this just sets all of the dates to the current date.

**Review by Crossedfall — CHANGES_REQUESTED** [review]

File: SQL/database_changelog.txt [hunk 4]
```diff
@@ -1,15 +1,43 @@
...
+CREATE TABLE `achievement_metadata` (
+	`achievement_key` VARCHAR(32) NOT NULL,
+	`achievement_version` SMALLINT UNSIGNED NOT NULL DEFAULT 0,
+	`achievement_type` enum('achievement','score','award') NULL DEFAULT NULL,
+	PRIMARY KEY (`achievement_key`)
+) ENGINE=InnoDB;
+
+Added achievement_name and achievement_description columns to achievement_metadata table.
+
+ALTER TABLE `achievement_metadata` ADD COLUMN (`achievement_name` VARCHAR(64) NULL DEFAULT NULL, `achievement_description` VARCHAR(512) NULL DEFAULT NULL);
```
Comment (Crossedfall) [inline SQL/database_changelog.txt]:
```suggestion
DROP TABLE IF EXISTS `achievement_metadata`;
CREATE TABLE `achievement_metadata` (
	`achievement_key` VARCHAR(32) NOT NULL,
	`achievement_version` SMALLINT UNSIGNED NOT NULL DEFAULT 0,
	`achievement_type` enum('achievement','score','award') NULL DEFAULT NULL,
	`achievement_name` VARCHAR(64) NULL DEFAULT NULL,
	`achievement_description` VARCHAR(512) NULL DEFAULT NULL,
	PRIMARY KEY (`achievement_key`)
) ENGINE=InnoDB;
```

There's no reason to have two queries when the table doesn't exist yet.

**Review by Crossedfall — CHANGES_REQUESTED** [review]

File: SQL/database_changelog.txt [hunk 5]
```diff
@@ -1,15 +1,41 @@
...
 
 In any query remember to add a prefix to the table names if you use one.
 
 -----------------------------------------------------
 
+Version 5.9, 5 October 2019 by Anturke, ported 1 October 2021 by Ivniinvi
+Added achievements table.
+See hub migration verb in _achievement_data.dm for details on migrating.
+
+CREATE TABLE `achievements` (
```
Comment (Crossedfall) [inline SQL/database_changelog.txt:18]:
```suggestion
DROP TABLE IF EXISTS `achievements`;
CREATE TABLE `achievements` (
```

Comment (Crossedfall) [reply SQL/database_changelog.txt:18]: Consistency with the other query

**Review by Crossedfall — APPROVED** [review]
Everything works now.

**Comment by Crossedfall** [comment]
DB updated and cheevos ported.

**Pull request merged**